### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by BaniaFonseca**
HI there

I added logging to api

please, see what kind of info should the api being logging

### CODE REVIEW

File: gunicorn.bash
Comment (tjiagoM) [inline gunicorn.bash]: What is the default value here? Shouldn't we have something specific here even if it's the same as the default one?

File: microservices/frame_engine/cameraCV.py
Comment (tjiagoM) [inline microservices/frame_engine/cameraCV.py]: this else clause is unnecessary, right?

File: microservices/object_detection/config.py
Comment (tjiagoM) [inline microservices/object_detection/config.py]: Shouldn't this be in another file? I don't think it's a good idea to include statements that are executed in a config module. Config file should only have defining variables/functions?

 Any reason to do that here though?

(p.s.: I have very limited knowledge on how module imports like the ones you do here work in practice with python, but I think I should still raise this issue)

File: microservices/object_detection/detector.py
Comment (tjiagoM) [inline microservices/object_detection/detector.py]: We should probably log some information from variable `e` to get more insights. Do we know the type of possible exceptions we can get?

File: microservices/object_detection/detector.py
Comment (tjiagoM) [inline microservices/object_detection/detector.py:30]: I think one possibility for logging is to calculate the time taken to load image, and get annotation. That will give us an idea if one part is taking too much time in separate. So we can check our bottlenecks per single image on average by only using the `time` module to calculate the time difference.

I'm not fully aware of how the `logging` module work, but it would be useful to have a file which we could somehow parse with these values so we could plot or create tables to gain insights.

File: microservices/object_detection/detector_base.py
Comment (tjiagoM) [inline microservices/object_detection/detector_base.py]: Are you sure that by loading this variable like this, then the rest of the code in `config.py` after `log = logging.getLogger()` is actually executed? I admit I have limited knowledge on how these things work with python imports, but doesn't seem obvious to me, I'd doublecheck or just organise things differently.

File: requirements.txt
Comment (tjiagoM) [inline requirements.txt]: Typos: We don't need **to** specify the .... **latest** version

File: setup.bash
Comment (tjiagoM) [inline setup.bash:49]: You define the alias before for `python` to be version 3, so you don't need to use `python3`

File: setup.bash
Comment (tjiagoM) [inline setup.bash]: Shouldn't you use the `version` variable you define in line 14 inside the echo? Otherwise changes in this variable won't have effect here.

File: setup.bash
Comment (tjiagoM) [inline setup.bash:22]: The `if $1` in line one is a bit odd. I think a comment line at the beginning of what this means could be helpful

**Review by tjiagoM — COMMENTED** [review]
Left a few questions.  My biggest confusion, maybe because of lack of knowledge is on how you setup the logging in the config module...

In terms of things to log, I suggested one thing for now in one of the comments.

File: gunicorn.bash
Comment (BaniaFonseca) [reply gunicorn.bash]: we are not going log with gunicorn, I'll remove that option in the future

File: microservices/object_detection/config.py
Comment (BaniaFonseca) [reply microservices/object_detection/config.py]: I needed the entire api use the same logger, but I can create a separated module that can be imported by other parts of the api that need to log

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py]: the possible types of exceptions will depend on the inputs to the api, and some other random events \
that's why I used  ``` log.exception ```, it will log the message and the exception, is not that cool?

File: setup.bash
Comment (EmersonAlvaro) [reply setup.bash:49]: Yes we do need to use python3, because if we execute make deploy=false setup the alias will not be executed and by default pip is execute with python2 so using python3 we are ensure that the python3 is being used.

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py:30]: got it, but can you elaborate more on that?\

I did not understand the calculation part, like, which sessions should we calculate?\ 
time need to load the image, time to do the detection, time to build the pascal voc xml, etc 

I can create a log sinilar to a csv file, so we can do the analysis later with pandas ...

File: microservices/object_detection/detector_base.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector_base.py]: it works perfectly, but I am going to create a module logger with a class on it, so that we can reuse it\

doing so we can have multiple loggers with differents porposes  ...

File: setup.bash
Comment (EmersonAlvaro) [reply setup.bash]: the folder is named python3.7 and not python3.7.3, and we use the variable version to get the binary file from source

File: setup.bash
Comment (EmersonAlvaro) [reply setup.bash:22]: Good ideia

Comment (BaniaFonseca) [reply setup.bash:22]: resolved

File: microservices/object_detection/config.py
Comment (tjiagoM) [reply microservices/object_detection/config.py]: Cool, I think in this case that change will be useful before this PR is merged :) (when you have the time, ofc)

File: microservices/object_detection/detector.py
Comment (tjiagoM) [reply microservices/object_detection/detector.py]: What I mean is, if there is an exception, the only thing we will know is "could not make detections". But we have more information on that in the `e` variable: eg what type of exception, the trace tree, maybe some more specific error. Therefore, I'm saying we should take that information and save it, so we can understand better where the exception is coming from (from `Annotator`, `Image`, etc, ?)

File: microservices/object_detection/detector.py
Comment (tjiagoM) [reply microservices/object_detection/detector.py:30]: Ah, sorry, the calculation part is just how usually these times are calculated. Usually is something similar to:

```python
t0 = get_timetep()
# Do calculation 
t1 = get_timestep()

total_time_taken = t1 - t0 # this is the calculation I was mentioning.
```

So, I'd say we could make these calculations in separate for loading the image, detection, building the xml, etc etc.

If we could have a csv file with these values it would be extremely useful to facilitate plot creation and further analysis :)

File: microservices/object_detection/detector_base.py
Comment (tjiagoM) [reply microservices/object_detection/detector_base.py]: Ok, in that case my concerns were nonsense. I need to read a bit more how the loading is done, given it turns out I had a wrong idea. However, as you say, it will be much better to have a reusable class!

File: setup.bash
Comment (tjiagoM) [reply setup.bash:49]: But that will be in lines 49-57, right? (after the `else`) Here you defined the python alias so you want to use the actual alias you defined a few lines above?

File: setup.bash
Comment (tjiagoM) [reply setup.bash]: Yes, but if you change `version` file in line 18 (line 14 in this PR), for example to 3.8.X, you would be installing that other  version but your alias (and path's python) would still be python3.7 that doesn't correspond to what you downloaded? Because in alias you are giving a binary, not a folder

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py]: I already said that ``` log.exception ```, will do the all work, both the message and the exception will be logged ...

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py:30]: got it

File: microservices/object_detection/detector.py
Comment (tjiagoM) [reply microservices/object_detection/detector.py]: Ahhhhhh! Now I got it, that's awesome :o I've just checked logging's documentation, I was not aware of that, very cool!

Comment (tjiagoM) [reply microservices/object_detection/detector.py]: You indeed said it before but I guess my bias didn't allow me to see the fact :D

File: setup.bash
Comment (EmersonAlvaro) [reply setup.bash]: The alias point to the folder which python where installed, and no to the binary  downloaded, if we change the version for 3.8.X for example we need to change the alias to point to folder where the new version(3.8) was been installed

File: setup.bash
Comment (EmersonAlvaro) [reply setup.bash:49]: Yes, but sometimes when we don't specify python like python3, we end up to use python2 and this is the best way to ensure that python3 is been used.

File: setup.bash
Comment (tjiagoM) [reply setup.bash]: I see. I think it's strange you download a Python-$version.tgz, but then regardless of what you download, you still go to python3.7...
I think it's very easy to have an unwanted effect here: someone changes $VERSION and then doesn't change the rest. Why not create a $VERSION_SHORT so the alias and the path are always used for the version you install?

All I'm saying is, given you have a general variable to select the python you download, I think you would have it complete by also specifying the alias/path to the version you install automatically.

File: microservices/frame_engine/cameraCV.py
Comment (nlauchande) [inline microservices/frame_engine/cameraCV.py]: I think we should have some time perf in here as well.

File: microservices/object_detection/detector.py
Comment (nlauchande) [inline microservices/object_detection/detector.py]: This multiplication *1000 everywhere reduces the readability of the code. The print function that you have down below can take care of conversion or formating.

File: microservices/object_detection/detector.py
Comment (nlauchande) [inline microservices/object_detection/detector.py:30]: There is much more instrumentation code than the actual functionality. I think it might be better to decouple this code. One approach to do that is with the use of timing decorators. Some ideas in here:  https://medium.com/pythonhive/python-decorator-to-measure-the-execution-time-of-methods-fa04cb6bb36d

**Review by nlauchande — CHANGES_REQUESTED** [review]
Some comments around readability of the methods with too much perf timing code affecting readability IMHO.

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py:30]: solved

File: microservices/object_detection/detector.py
Comment (BaniaFonseca) [reply microservices/object_detection/detector.py]: solved

**Review by tjiagoM — COMMENTED** [review]
I checked the new files since my last comments and didn't spot anything "new" besides what I've already said and Natu mentioned.

I'd say we should be careful with the PDF files so that they don't occupy too much space. 

And maybe next time we can start naming the PR (rather than just having "Dev"), so it has a semantic and helps us keep it more focused :)

File: scripts/setup.sh
Comment (tjiagoM) [inline scripts/setup.sh]: two typos: for more *details* (plural) refer to *README.md*

**Review by nlauchande — APPROVED** [review]

**Review by tjiagoM — APPROVED** [review]
I think we can merge this (though after @nlauchande sees his requested changes). 

I like the new organisation inside `object_detection`!

I continue to say that it would be a better idea to specify the versions in the `requirements.txt`. Installing the latest version it's not a good idea because sometimes it breaks incompatibility (what if this happens when we are about to deploy the pi into the observatory?). This is probably related to issue #4, and if this issue is the reason to stop us to specify versions we should probably have a look.

**Pull request merged**
